File: wizards/com/sun/star/wizards/table/TableWizard.java
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This file incorporates work covered by the following license notice:
 *
 *   Licensed to the Apache Software Foundation (ASF) under one or more
 *   contributor license agreements. See the NOTICE file distributed
 *   with this work for additional information regarding copyright
 *   ownership. The ASF licenses this file to you under the Apache
 *   License, Version 2.0 (the "License"); you may not use this file
 *   except in compliance with the License. You may obtain a copy of
 *   the License at http://www.apache.org/licenses/LICENSE-2.0 .
 */
package com.sun.star.wizards.table;

import java.util.HashMap;

import com.sun.star.awt.TextEvent;
import com.sun.star.awt.VclWindowPeerAttribute;
import com.sun.star.awt.XTextListener;
import com.sun.star.beans.PropertyValue;
import com.sun.star.lang.XInitialization;
import com.sun.star.lang.XMultiServiceFactory;
import com.sun.star.sdb.CommandType;
import com.sun.star.sdb.application.DatabaseObject;
import com.sun.star.sdbc.SQLException;
import com.sun.star.task.XJobExecutor;
import com.sun.star.uno.UnoRuntime;
import com.sun.star.wizards.common.*;
import com.sun.star.wizards.db.DatabaseObjectWizard;
import com.sun.star.wizards.db.TableDescriptor;
import com.sun.star.wizards.ui.*;

public class TableWizard extends DatabaseObjectWizard implements XTextListener
{

    private static String slblFields;
    private static String slblSelFields;
    private Finalizer curFinalizer;
    private ScenarioSelector curScenarioSelector;
    private FieldFormatter curFieldFormatter;
    private PrimaryKeyHandler curPrimaryKeyHandler;
    HashMap<String, FieldDescription> fielditems;
    int wizardmode;
    private String tablename;
    private String serrToManyFields;
    private String serrTableNameexists;
    private String scomposedtablename;
    private TableDescriptor curTableDescriptor;
    public static final int SOMAINPAGE = 1;
    public static final int SOFIELDSFORMATPAGE = 2;
    public static final int SOPRIMARYKEYPAGE = 3;
    public static final int SOFINALPAGE = 4;
    private String sMsgColumnAlreadyExists = PropertyNames.EMPTY_STRING;

    private String m_tableName;

    public TableWizard( XMultiServiceFactory xMSF, PropertyValue[] i_wizardContext )
    {
        super( xMSF, 41200, i_wizardContext );
        super.addResourceHandler("dbw");
        String sTitle = m_oResource.getResText(UIConsts.RID_TABLE + 1);
        Helper.setUnoPropertyValues(xDialogModel,
                new String[]
                {
                    PropertyNames.PROPERTY_HEIGHT, PropertyNames.PROPERTY_MOVEABLE, PropertyNames.PROPERTY_NAME, PropertyNames.PROPERTY_POSITION_X, PropertyNames.PROPERTY_POSITION_Y, PropertyNames.PROPERTY_STEP, PropertyNames.PROPERTY_TABINDEX, PropertyNames.PROPERTY_TITLE, PropertyNames.PROPERTY_WIDTH
                },
                new Object[]
                {
                    218, Boolean.TRUE, "DialogTable", 102, 41, 1, Short.valueOf((short) 0), sTitle, 330
                });
        drawNaviBar();
        fielditems = new HashMap<String, FieldDescription>();
        //TODO if reportResources cannot be gotten dispose officedocument
        if (getTableResources())
        {
            setRightPaneHeaders(m_oResource, UIConsts.RID_TABLE + 8, 4);
        }
    }

    @Override
    protected void leaveStep(int nOldStep, int nNewStep)
    {
        switch (nOldStep)
        {
            case SOMAINPAGE:
                curScenarioSelector.addColumnsToDescriptor();
                break;
            case SOFIELDSFORMATPAGE:
                curFieldFormatter.updateColumnofColumnDescriptor();
                String[] sfieldnames = curFieldFormatter.getFieldNames();
                super.setStepEnabled(SOFIELDSFORMATPAGE, sfieldnames.length > 0);
                curScenarioSelector.setSelectedFieldNames(sfieldnames);
                break;
            case SOPRIMARYKEYPAGE:
                break;
            case SOFINALPAGE:
                break;
            default:
                break;
        }
    }

    @Override
    protected void enterStep(int nOldStep, int nNewStep)
    {
        switch (nNewStep)
        {
            case SOMAINPAGE:
                break;
            case SOFIELDSFORMATPAGE:
                curFieldFormatter.initialize(curTableDescriptor, this.curScenarioSelector.getSelectedFieldNames());
                break;
            case SOPRIMARYKEYPAGE:
                curPrimaryKeyHandler.initialize();
                break;
            case SOFINALPAGE:
                curFinalizer.initialize(curScenarioSelector.getFirstTableName());
                break;
            default:
                break;
        }
    }


    public boolean iscompleted(int _ndialogpage)
    {
        switch (_ndialogpage)
        {
            case SOMAINPAGE:
                return curScenarioSelector.iscompleted();
            case SOFIELDSFORMATPAGE:
                return this.curFieldFormatter.iscompleted();
            case SOPRIMARYKEYPAGE:
                if (curPrimaryKeyHandler != null)
                {
                    return this.curPrimaryKeyHandler.iscompleted();
                }
            case SOFINALPAGE:
                return this.curFinalizer.iscompleted();
            default:
                return false;
        }
    }


    public void setcompleted(int _ndialogpage, boolean _biscompleted)
    {
        boolean bScenarioiscompleted = _biscompleted;
        boolean bPrimaryKeysiscompleted = _biscompleted;
        boolean bFinalPageiscompleted = _biscompleted;
        if (_ndialogpage == SOMAINPAGE)
        {
            curFinalizer.initialize(curScenarioSelector.getFirstTableName());
        }
        else
        {
            bScenarioiscompleted = iscompleted(SOMAINPAGE);
        }
        if (_ndialogpage != TableWizard.SOPRIMARYKEYPAGE && (this.curPrimaryKeyHandler != null))
        {
            bPrimaryKeysiscompleted = iscompleted(SOPRIMARYKEYPAGE);
        }
        if (_ndialogpage != TableWizard.SOFINALPAGE)
        {
            bFinalPageiscompleted = iscompleted(SOFINALPAGE);           // Basically the finalpage is always enabled
        }
        if (bScenarioiscompleted)
        {
            super.setStepEnabled(SOFIELDSFORMATPAGE, true);
            super.setStepEnabled(SOPRIMARYKEYPAGE, true);
            if (bPrimaryKeysiscompleted)
            {
                super.enablefromStep(SOFINALPAGE, true);
                super.enableFinishButton(bFinalPageiscompleted);
            }
            else
            {
                super.enablefromStep(SOFINALPAGE, false);
                enableNextButton(false);
            }
        }
        else if (_ndialogpage == SOFIELDSFORMATPAGE)
        {
            super.enablefromStep(super.getCurrentStep() + 1, iscompleted(SOFIELDSFORMATPAGE));
        }
        else
        {
            super.enablefromStep(super.getCurrentStep() + 1, false);
        }
    }

/*
    public static void main(String args[])
    {
        String ConnectStr = "uno:socket,host=localhost,port=8100;urp,negotiate=0,forcesynchronous=1;StarOffice.NamingService";
        PropertyValue[] curproperties = null;
        try
        {
            XMultiServiceFactory xLocMSF = com.sun.star.wizards.common.Desktop.connect(ConnectStr);
            TableWizard CurTableWizard = new TableWizard(xLocMSF);
            if (xLocMSF != null)
            {
                System.out.println("Connected to " + ConnectStr);
                curproperties = new PropertyValue[1];
                curproperties[0] = Properties.createProperty("DataSourceName", "Bibliography");
                //curproperties[0] = Properties.createProperty("DatabaseLocation", "file:///path/to/database.odb");
                CurTableWizard.startTableWizard(xLocMSF, curproperties);
            }
        }
        catch (Exception exception)
        {
            exception.printStackTrace(System.err);
        }
    }
*/
    private void buildSteps()
    {
        curScenarioSelector = new ScenarioSelector(this, this.curTableDescriptor, slblFields, slblSelFields);
        curFieldFormatter = new FieldFormatter(this);
        if ( this.curTableDescriptor.supportsPrimaryKeys() )
        {
            curPrimaryKeyHandler = new PrimaryKeyHandler(this, curTableDescriptor);
        }
        curFinalizer = new Finalizer(this, curTableDescriptor);
        enableNavigationButtons(false, false, false);
    }

    private boolean createTable()
    {
        boolean bIsSuccessfull = true;
        boolean bTableCreated = false;
        String schemaname = curFinalizer.getSchemaName();
        String catalogname = curFinalizer.getCatalogName();
        if (curTableDescriptor.supportsPrimaryKeys())
        {
            String[] keyfieldnames = curPrimaryKeyHandler.getPrimaryKeyFields();
            if (keyfieldnames != null && keyfieldnames.length > 0)
            {
                boolean bIsAutoIncrement = curPrimaryKeyHandler.isAutoIncremented();
                bIsSuccessfull = curTableDescriptor.createTable(catalogname, schemaname, tablename, keyfieldnames, bIsAutoIncrement);
                bTableCreated = true;
            }
        }
        if (!bTableCreated)
        {
            bIsSuccessfull = curTableDescriptor.createTable(catalogname, schemaname, tablename);
        }
        if ((!bIsSuccessfull) && (curPrimaryKeyHandler.isAutomaticMode()))
        {
            curTableDescriptor.dropColumnbyName(curPrimaryKeyHandler.getAutomaticFieldName());
        }
        return bIsSuccessfull;
    }

    @Override
    public boolean finishWizard()
    {
        super.switchToStep(super.getCurrentStep(), SOFINALPAGE);
        tablename = curFinalizer.getTableName(curScenarioSelector.getFirstTableName());
        scomposedtablename = curFinalizer.getComposedTableName(tablename);
        if (this.curTableDescriptor.isSQL92CheckEnabled())
        {
            Desktop.removeSpecialCharacters(curTableDescriptor.xMSF, Configuration.getLocale(this.curTableDescriptor.xMSF), tablename);
        }
        if ( tablename.length() > 0 )
        {
            if (!curTableDescriptor.hasTableByName(scomposedtablename))
            {
                wizardmode = curFinalizer.finish();
                if (createTable())
                {
                    final boolean editTableDesign = (wizardmode == Finalizer.MODIFYTABLEMODE );
                    loadSubComponent( DatabaseObject.TABLE, curTableDescriptor.getComposedTableName(), editTableDesign );
                    m_tableName = curTableDescriptor.getComposedTableName();
                    super.xDialog.endExecute();
                    return true;
                }
            }
            else
            {
                String smessage = JavaTools.replaceSubString(serrTableNameexists, tablename, "%TABLENAME");
                super.showMessageBox("WarningBox", com.sun.star.awt.VclWindowPeerAttribute.OK, smessage);
                curFinalizer.setFocusToTableNameControl();
            }
        }
        return false;
    }

    private void callFormWizard()
    {
        try
        {
            Object oFormWizard = this.xMSF.createInstance("com.sun.star.wizards.form.CallFormWizard");

            NamedValueCollection wizardContext = new NamedValueCollection();
            wizardContext.put( PropertyNames.ACTIVE_CONNECTION, curTableDescriptor.DBConnection );
            wizardContext.put( "DataSource", curTableDescriptor.getDataSource() );
            wizardContext.put( PropertyNames.COMMAND_TYPE, CommandType.TABLE );
            wizardContext.put( PropertyNames.COMMAND, scomposedtablename );
            wizardContext.put( "DocumentUI", m_docUI );
            XInitialization xInitialization = UnoRuntime.queryInterface( XInitialization.class, oFormWizard );
            xInitialization.initialize( wizardContext.getPropertyValues() );
            XJobExecutor xJobExecutor = UnoRuntime.queryInterface( XJobExecutor.class, oFormWizard );
            xJobExecutor.trigger(PropertyNames.START);
        }
        catch (Exception e)
        {
            e.printStackTrace(System.err);
        }
    }

    @Override
    public void cancelWizard()
    {
        xDialog.endExecute();
    }

    private void insertFormRelatedSteps()
    {
        addRoadmap();
        int i = 0;
        i = insertRoadmapItem(0, true, m_oResource.getResText(UIConsts.RID_TABLE + 2), SOMAINPAGE);
        i = insertRoadmapItem(i, false, m_oResource.getResText(UIConsts.RID_TABLE + 3), SOFIELDSFORMATPAGE);
        if (this.curTableDescriptor.supportsPrimaryKeys())
        {
            i = insertRoadmapItem(i, false, m_oResource.getResText(UIConsts.RID_TABLE + 4), SOPRIMARYKEYPAGE);
        }
        i = insertRoadmapItem(i, false, m_oResource.getResText(UIConsts.RID_TABLE + 5), SOFINALPAGE);        // Orderby is always supported
        setRoadmapInteractive(true);
        setRoadmapComplete(true);
        setCurrentRoadmapItemID((short) 1);
    }

    public String startTableWizard(  )
    {
        try
        {
            curTableDescriptor = new TableDescriptor(xMSF, super.xWindow, this.sMsgColumnAlreadyExists);
            if ( curTableDescriptor.getConnection( m_wizardContext ) )
            {
                buildSteps();
                createWindowPeer();
                curTableDescriptor.setWindowPeer(this.xControl.getPeer());
                insertFormRelatedSteps();
                short RetValue = executeDialog();
                xComponent.dispose();
                if ( RetValue == 0 )
                {
                    if (  wizardmode == Finalizer.STARTFORMWIZARDMODE )
                        callFormWizard();
                    return m_tableName;
                }
            }
        }
        catch (java.lang.Exception jexception)
        {
            jexception.printStackTrace(System.err);
        }
        return PropertyNames.EMPTY_STRING;
    }

    private boolean getTableResources()
    {
        super.m_oResource.getResText(UIConsts.RID_TABLE + 1);
        slblFields = m_oResource.getResText(UIConsts.RID_TABLE + 19);
        slblSelFields = m_oResource.getResText(UIConsts.RID_TABLE + 25);
        serrToManyFields = m_oResource.getResText(UIConsts.RID_TABLE + 47);
        serrTableNameexists = m_oResource.getResText(UIConsts.RID_TABLE + 48);
        sMsgColumnAlreadyExists = m_oResource.getResText(UIConsts.RID_TABLE + 51);
        return true;
    }

    public boolean verifyfieldcount(int _icount)
    {
        try
        {
            int maxfieldcount = curTableDescriptor.getMaxColumnsInTable();
            if (_icount >= (maxfieldcount - 1))
            {   // keep one column as reserve for the automatically created key
                String smessage = serrToManyFields;
                smessage = JavaTools.replaceSubString(smessage, String.valueOf(maxfieldcount), "%COUNT");
                showMessageBox("ErrorBox", VclWindowPeerAttribute.OK, smessage);
                return false;
            }
        }
        catch (SQLException e)
        {
            e.printStackTrace(System.err);
        }
        return true;
    }


    /* (non-Javadoc)
     * @see com.sun.star.awt.XTextListener#textChanged(com.sun.star.awt.TextEvent)
     */
    public void textChanged(TextEvent aTextEvent)
    {
        if (this.curTableDescriptor.isSQL92CheckEnabled())
        {
            Object otextcomponent = UnoDialog.getModel(aTextEvent.Source);
            String sName = (String) Helper.getUnoPropertyValue(otextcomponent, "Text");
            sName = Desktop.removeSpecialCharacters(curTableDescriptor.xMSF, Configuration.getLocale(curTableDescriptor.xMSF), sName);
            Helper.setUnoPropertyValue(otextcomponent, "Text", sName);
        }
    }
}

File: wizards/com/sun/star/wizards/table/TableWizard.py
#
# This file is part of the LibreOffice project.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
# This file incorporates work covered by the following license notice:
#
#   Licensed to the Apache Software Foundation (ASF) under one or more
#   contributor license agreements. See the NOTICE file distributed
#   with this work for additional information regarding copyright
#   ownership. The ASF licenses this file to you under the Apache
#   License, Version 2.0 (the "License"); you may not use this file
#   except in compliance with the License. You may obtain a copy of
#   the License at http://www.apache.org/licenses/LICENSE-2.0 .
#
from .TableWizardResources import TableWizardResources
from ..ui.WizardDialog import WizardDialog, uno, PropertyNames
from ..db.TableDescriptor import TableDescriptor
from ..common.Desktop import Desktop

class TableWizard(WizardDialog):
    HID = 41200
    SONULLPAGE = 0
    SOMAINPAGE = 1
    SOFIELDSFORMATPAGE = 2
    SOPRIMARYKEYPAGE = 3
    SOFINALPAGE = 4

    def __init__(self, xmsf):       
        super(TableWizard,self).__init__(xmsf, TableWizard.HID)
        slblFields = str()
        slblSelFields = str()
        wizardmode = int()
        tablename = str()
        serrToManyFields = str()
        serrTableNameexists = str()
        scomposedtablename = str()
        m_tableName = str()
    
    def startWizard(self, xmsf):

        #Load Resources
        self.resources = TableWizardResources(self.oWizardResource)
                        
        #set dialog properties...
        self.setDialogProperties(True, 218, True, 102, 41, 1, 0, 
                self.resources.sTitle, 330)              
                
        self.drawNaviBar()
        self.setRightPaneHeaders(self.resources.RoadmapLabels)
        
        self.curTableDescriptor = TableDescriptor(xmsf, self.sMsgColumnAlreadyExists)
        ## TODO-XISCO if self.curTableDescriptor.getConnection(m_wizardContext):
        #self.buildSteps()
        self.createWindowPeer()
        self.curTableDescriptor.setWindowPeer(self.xUnoDialog.getPeer())
        self.insertFormRelatedSteps()
        xComponent.dispose()
        if RetValue == 0:
            if self.wizardmode == Finalizer.STARTFORMWIZARDMODE:
                self.callFormWizard()
            return self.m_tableName        
                   
    def leaveStep(self, nOldStep, nNewStep):
        if nOldStep==self.SOMAINPAGE:
            self.curScenarioSelector.addColumnsToDescriptor()
        elif nOldStep==self.SOFIELDSFORMATPAGE:
            self.curFieldFormatter.updateColumnofColumnDescriptor()
            super(TableWizard, self).setStepEnabled(self.SOFIELDSFORMATPAGE, len(sfieldnames))
            self.curScenarioSelector.setSelectedFieldNames(sfieldnames)

    def enterStep(self, nOldStep, nNewStep):
        if nNewStep==self.SOFIELDSFORMATPAGE:
            self.curFieldFormatter.initialize(self.curTableDescriptor, self.curScenarioSelector.getSelectedFieldNames())
        elif nNewStep==self.SOPRIMARYKEYPAGE:
            self.curPrimaryKeyHandler.initialize()
        elif nNewStep==self.SOFINALPAGE:
            self.curFinalizer.initialize(self.curScenarioSelector.getFirstTableName())
   
    def iscompleted(self, _ndialogpage):
        if _ndialogpage==self.SOMAINPAGE:
            return self.curScenarioSelector.iscompleted()
        elif _ndialogpage==self.SOFIELDSFORMATPAGE:
            return self.curFieldFormatter.iscompleted()
        elif _ndialogpage==self.SOPRIMARYKEYPAGE:
            if self.curPrimaryKeyHandler != None:
                return self.curPrimaryKeyHandler.iscompleted()
        elif _ndialogpage==self.SOFINALPAGE:
            return self.curFinalizer.iscompleted()
        else:
            return False
   
    def setcompleted(self, _ndialogpage, _biscompleted):
        bScenarioiscompleted = _biscompleted
        bFieldFormatsiscompleted = _biscompleted
        bPrimaryKeysiscompleted = _biscompleted
        bFinalPageiscompleted = _biscompleted
        if _ndialogpage == self.SOMAINPAGE:
            self.curFinalizer.initialize(self.curScenarioSelector.getFirstTableName())
        else:
            bScenarioiscompleted = self.iscompleted(self.SOMAINPAGE)
        if _ndialogpage != TableWizard.SOFIELDSFORMATPAGE:
            bFieldFormatsiscompleted = self.iscompleted(self.SOFIELDSFORMATPAGE)
            if not bFieldFormatsiscompleted:
                #  it might be that the Fieldformatter has not yet been initialized
                bFieldFormatsiscompleted = bScenarioiscompleted
                #  in this case query the scenarioselector
        if _ndialogpage != TableWizard.SOPRIMARYKEYPAGE and (self.curPrimaryKeyHandler != None):
            bPrimaryKeysiscompleted = self.iscompleted(self.SOPRIMARYKEYPAGE)
        if _ndialogpage != TableWizard.SOFINALPAGE:
            bFinalPageiscompleted = self.iscompleted(self.SOFINALPAGE)
            #  Basically the finalpage is always enabled
        if bScenarioiscompleted:
            super(TableWizard, self).setStepEnabled(self.SOFIELDSFORMATPAGE, True)
            super(TableWizard, self).setStepEnabled(self.SOPRIMARYKEYPAGE, True)
            if bPrimaryKeysiscompleted:
                super(TableWizard, self).enablefromStep(self.SOFINALPAGE, True)
                super(TableWizard, self).enableFinishButton(bFinalPageiscompleted)
            else:
                super(TableWizard, self).enablefromStep(self.SOFINALPAGE, False)
                enableNextButton(False)
        elif _ndialogpage == self.SOFIELDSFORMATPAGE:
            super(TableWizard, self).enablefromStep(super(TableWizard, self).getCurrentStep() + 1, self.iscompleted(self.SOFIELDSFORMATPAGE))
        else:
            super(TableWizard, self).enablefromStep(super(TableWizard, self).getCurrentStep() + 1, False)

    def buildSteps(self):
        self.curScenarioSelector = ScenarioSelector(self, self.curTableDescriptor, self.slblFields, self.slblSelFields)
        self.curFieldFormatter = FieldFormatter(self, self.curTableDescriptor)
        if self.curTableDescriptor.supportsPrimaryKeys():
            self.curPrimaryKeyHandler = PrimaryKeyHandler(self, self.curTableDescriptor)
        self.curFinalizer = Finalizer(self, self.curTableDescriptor)
        enableNavigationButtons(False, False, False)

    def createTable(self):
        bIsSuccessfull = True
        bTableCreated = False
        schemaname = self.curFinalizer.getSchemaName()
        catalogname = self.curFinalizer.getCatalogName()
        if self.curTableDescriptor.supportsPrimaryKeys():
            if keyfieldnames != None:
                if len(keyfieldnames):
                    bIsSuccessfull = self.curTableDescriptor.createTable(catalogname, schemaname, self.tablename, keyfieldnames, bIsAutoIncrement, self.curScenarioSelector.getSelectedFieldNames())
                    bTableCreated = True
        if not bTableCreated:
            bIsSuccessfull = self.curTableDescriptor.createTable(catalogname, schemaname, self.tablename, self.curScenarioSelector.getSelectedFieldNames())
        if (not bIsSuccessfull) and (self.curPrimaryKeyHandler.isAutomaticMode()):
            self.curTableDescriptor.dropColumnbyName(self.curPrimaryKeyHandler.getAutomaticFieldName())
        return bIsSuccessfull

    def finishWizard(self):
        super(TableWizard, self).switchToStep(super(TableWizard, self).getCurrentStep(), self.SOFINALPAGE)
        self.tablename = self.curFinalizer.getTableName(self.curScenarioSelector.getFirstTableName())
        self.scomposedtablename = self.curFinalizer.getComposedTableName(self.tablename)
        if self.curTableDescriptor.isSQL92CheckEnabled():
            Desktop.removeSpecialCharacters(self.curTableDescriptor.xMSF, Configuration.getLocale(self.curTableDescriptor.xMSF), self.tablename)
        if 0 > len(tablename):
            if not self.curTableDescriptor.hasTableByName(self.scomposedtablename):
                self.wizardmode = self.curFinalizer.finish()
                if self.createTable():
                    loadSubComponent(DatabaseObject.TABLE, self.curTableDescriptor.getComposedTableName(), editTableDesign)
                    self.m_tableName = self.curTableDescriptor.getComposedTableName()
                    super(TableWizard, self).xDialog.endExecute()
                    return True
            else:
                super(TableWizard, self).showMessageBox("WarningBox", com.sun.star.awt.VclWindowPeerAttribute.OK, smessage)
                self.curFinalizer.setFocusToTableNameControl()
        return False

    def callFormWizard(self):
        try:
            wizardContext.put(PropertyNames.ACTIVE_CONNECTION, self.curTableDescriptor.DBConnection)
            wizardContext.put("DataSource", self.curTableDescriptor.getDataSource())
            wizardContext.put(PropertyNames.COMMAND_TYPE, CommandType.TABLE)
            wizardContext.put(PropertyNames.COMMAND, self.scomposedtablename)
            wizardContext.put("DocumentUI", m_docUI)
            xInitialization.initialize(wizardContext.getPropertyValues())
            xJobExecutor.trigger(PropertyNames.START)
        except Exception as e:
            e.printStackTrace(System.err)

    def cancelWizard(self):
        xDialog.endExecute()

    def insertFormRelatedSteps(self):
        self.addRoadmap()

        setRoadmapInteractive(True)
        setRoadmapComplete(True)
        setCurrentRoadmapItemID(int(1))

    def verifyfieldcount(self, _icount):
        try:
            if _icount >= (maxfieldcount - 1):
                #  keep one column as reserve for the automaticcally created key
                smessage = JavaTools.replaceSubString(smessage, String.valueOf(maxfieldcount), "%COUNT")
                showMessageBox("ErrorBox", VclWindowPeerAttribute.OK, smessage)
                return False
        except SQLException as e:
            e.printStackTrace(System.err)
        return True
   
    def textChanged(self, aTextEvent):
        if self.curTableDescriptor.isSQL92CheckEnabled():
            sName = Desktop.removeSpecialCharacters(self.curTableDescriptor.xMSF, Configuration.getLocale(self.curTableDescriptor.xMSF), sName)
            Helper.setUnoPropertyValue(otextcomponent, "Text", sName)
